llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -505,13 +548,15 @@ class VSETVLIInfo {
   bool getTailAgnostic() const { return TailAgnostic; }
   bool getMaskAgnostic() const { return MaskAgnostic; }
 
-  bool hasNonZeroAVL(const MachineRegisterInfo &MRI) const {
+  bool hasNonZeroAVL(const MachineRegisterInfo &MRI,
+                     const LiveIntervals *LIS) const {
     if (hasAVLImm())
       return getAVLImm() > 0;
     if (hasAVLReg()) {
       if (getAVLReg() == RISCV::X0)
         return true;
-      if (MachineInstr *MI = MRI.getVRegDef(getAVLReg());
+      if (MachineInstr *MI =
+              getReachingDefMI(getAVLReg(), (MachineInstr *)nullptr, &MRI, LIS);
```

Comment:
If we're passing nullptr to MI then won't this will always fail if run post reg alloc? Maybe we need to store the MachineInstr that defines the AVL instead of the register in VSETVLIInfo

Reply:
You're right. Keeping the AVLReg define MI inside `VSETVLIInfo` will give a better result. I'll implement it. Thanks!

Reply:
Create a pr to keep the AVLRegDef inside VSETVLInfo. https://github.com/llvm/llvm-project/pull/89180